collector/src/main/java/io/prometheus/jmx/JmxScraper.java

```diff
@@ -135,21 +137,31 @@ private void scrapeBean(MBeanServerConnection beanConn, ObjectName mbeanName) {
         }
         MBeanAttributeInfo[] attrInfos = info.getAttributes();
 
+        Map<String, MBeanAttributeInfo> name2AttrInfo = new LinkedHashMap<String, MBeanAttributeInfo>();
         for (int idx = 0; idx < attrInfos.length; ++idx) {
             MBeanAttributeInfo attr = attrInfos[idx];
             if (!attr.isReadable()) {
                 logScrape(mbeanName, attr, "not readable");
                 continue;
             }
-
-            Object value;
-            try {
-                value = beanConn.getAttribute(mbeanName, attr.getName());
-            } catch(Exception e) {
-                logScrape(mbeanName, attr, "Fail: " + e);
-                continue;
+            name2AttrInfo.put(attr.getName(), attr);
+        }
+        AttributeList attributes = new AttributeList();
+        try {
+            attributes = beanConn.getAttributes(mbeanName, name2AttrInfo.keySet().toArray(new String[0]));
+        } catch (Exception e) {
+            //couldn't get them all in one go, try them 1 by 1
```

Comment:
Why do you need this?

Reply:
following your review comment
"Handle things like this gracefully, many JMX servers are broken and attributes can change"
...rather than throw the exception to say I couldn't get any attributes
I thought why not try them individually (just incase one attribute is adversely affecting the others - ie revert back to the original behaviour) on the up side you may get at least some of the attributes back on the down side you will get a longer list of error messages

Reply:
That's not the failure mode I was worried about, you can just log here.

Reply:
do you mean - don't try the one at a time approach - just log the failure and continue?

Reply:
Yes

Reply:
ok - will be able to check that in a bit later